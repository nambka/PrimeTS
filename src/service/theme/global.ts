import { createGlobalStyle } from 'styled-components';

// https://lewislbr.dev/blog/add-dark-mode-react-typescript-styled-components/
// https://blog.ezrabowman.com/build-a-theme-switcher-in-react-using-styled-components-and-redux/
// https://medium.com/rbi-tech/theme-with-styled-components-and-typescript-209244ec15a3

export const GlobalStyle = createGlobalStyle`
  *,
  *::after,
  *::before {
    box-sizing: border-box;
  }
  
  body {
    /* background: ${(props: any) => props.theme.bodyBkg}; */
    background: ${({ theme }) => theme.bodyBkg};
    color: ${({ theme }) => theme.bodyTextColor};
    font-family: BlinkMacSystemFont, -apple-system, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif;
    transition: all 0.25s linear;
  }

  footer {
    background: ${({ theme }) => theme.footerBkg};
    color: ${({ theme }) => theme.footerTextColor};
  }

  form {
    background-color: ${({ theme }) => theme.formBkg};
    border: ${({ theme }) => theme.formBorder};
    box-shadow: ${({ theme }) => theme.formBoxShadow};
    transition: ${({ theme }) => theme.formTransition};
  }

  form:hover {
    background-color: ${({ theme }) => theme.formBkgHover};
    border: ${({ theme }) => theme.formBorder};
    box-shadow: ${({ theme }) => theme.formBoxShadowHover};
    transition: ${({ theme }) => theme.formTransitionHover};
  }

  form .nambk-header-desc {
    color: ${({ theme }) => theme.formHeaderDescColor};
  }

  input[type=number] {
    color: ${({ theme }) => theme.bodyTextColor};
    background-color: ${({ theme }) => theme.inputBkg};
    border: ${({ theme }) => theme.inputBorder};
  }

  input[type=number]:focus {
    border-color: ${({ theme }) => theme.inputBorderHover};
  }

  #result {
    box-shadow: ${({ theme }) => theme.resultBoxShadow};
  }
`
// export default GlobalStyle;