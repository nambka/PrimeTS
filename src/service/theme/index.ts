import { useState, useEffect } from 'react';

export const useDarkMode = () => {
    const [theme, setTheme] = useState(localStorage.getItem('theme') || 'light');
    const toggleTheme = () => {
      // theme === 'light' ? setTheme('dark') : setTheme('light')
      if (theme === 'light') {
        localStorage.setItem('theme', 'dark');
        setTheme('dark');
      } else if (theme === 'dark') {
        localStorage.setItem('theme', 'moon');
        setTheme('moon');
      } else {
        localStorage.setItem('theme', 'light');
        setTheme('light');
      }
    };

    useEffect(() => {
      const localTheme = window.localStorage.getItem('theme');
      localTheme && setTheme(localTheme);
    }, []);
    
    return [theme, toggleTheme];
};

// declare module 'styled-components' {
//   type Theme = typeof lightTheme;
//   export interface DefaultTheme extends Theme {}
// }

export const lightTheme = {
  bodyBkg: '#F7F8F9',
  bodyTextColor: '#353637',
  footerBkg: 'linear-gradient(150deg,#19a0ff,#2d1582)',
  footerTextColor: '#fff',

  formBorder: '1px solid #eaecef22',

  formBkg: '#fff!important',
  // formBoxShadow: '0 2px 2px 0 rgba(0,0,0,.05)',
  formBoxShadow: '0 5px 30px 0 rgba(0,0,0,.2)',
  formTransition: '.3s all ease',

  formBkgHover: '#fff',
  formBoxShadowHover: '0 3px 10px 0 rgba(204,204,204,.8)',
  formTransitionHover: '.2s all ease',

  formHeaderDescColor: '#006647', // Contrast Checker:https://webaim.org/resources/contrastchecker/

  inputBkg: '#fff3',
  inputBorder: '1px solid #E6E8EA',
  inputBorderHover: '#F0B90B',

  resultBoxShadow: 'inset 0 0 5px 0 rgba(0,0,0,.1)', 
}

export const darkTheme = {
  bodyBkg: '#111',
  bodyTextColor: '#ccc',
  footerBkg: '#000',
  footerTextColor: '#ccc',

  formBorder: '1px solid #aaa3',

  formBkg: '#000!important',
  formBoxShadow: '0 0.5rem 1.2rem rgba(124,124,124,.2)',
  formTransition: '.3s all ease',

  formBkgHover: '#bbb1!important',
  formBoxShadowHover: '2px 5px 10px 0 rgba(0,0,0,.2)',
  formTransitionHover: '.2s all ease',

  formHeaderDescColor: '#00c087', // Contrast Checker:https://webaim.org/resources/contrastchecker/
  
  inputBkg: '#ccc2',
  inputBorder: '1px solid #555',
  inputBorderHover: '#1274e0',

  resultBoxShadow: 'inset 0 0 5px 0 rgba(0,0,0,.3)',
}

export const moonTheme = {
  bodyBkg: '#01263f',
  bodyTextColor: '#AABCCB',
  footerBkg: '#012137',
  footerTextColor: '#AABCCB',

  formBorder: '1px solid #18365b',

  formBkg: '#012137!important',
  // formBoxShadow: '0 2px 2px 0 rgba(0,0,0,.1)',
  formBoxShadow: '0 0.5rem 1.2rem rgba(4,76,124,.2)',
  formTransition: '.3s all ease',

  formBkgHover: '#bbb1!important',
  formBoxShadowHover: '2px 5px 10px 0 rgba(0,0,0,.3)',
  // formBoxShadowHover: '0 0.5rem 1.2rem rgba(4,76,124,.2)',
  formTransitionHover: '.2s all ease',

  formHeaderDescColor: '#00D696', // Contrast Checker:https://webaim.org/resources/contrastchecker/
  
  inputBkg: '#ccd2',
  inputBorder: '1px solid #256',
  inputBorderHover: '#2ac',

  resultBoxShadow: 'inset 0 0 5px 0 rgba(0,0,0,.3)',
  // resultBoxShadow: 'inset 0 0 10px #ddd;',
}

// const themes = {
//   lightTheme: {
//     bodyBkg: '#F7F8F9',
//     bodyTextColor: '#353637',
//     footerBkg: 'linear-gradient(150deg,#19a0ff,#2d1582)',
//     footerTextColor: '#fff',

//     formBorder: '1px solid #eaecef22',

//     formBkg: '#fff!important',
//     // formBoxShadow: '0 2px 2px 0 rgba(0,0,0,.05)',
//     formBoxShadow: '0 5px 30px 0 rgba(0,0,0,.2)',
//     formTransition: '.3s all ease',

//     formBkgHover: '#fff',
//     formBoxShadowHover: '0 3px 10px 0 rgba(204,204,204,.8)',
//     formTransitionHover: '.2s all ease',

//     formHeaderDescColor: '#006647', // Contrast Checker:https://webaim.org/resources/contrastchecker/

//     inputBkg: '#fff3',
//     inputBorder: '1px solid #E6E8EA',
//     inputBorderHover: '#F0B90B',

//     resultBoxShadow: 'inset 0 0 5px 0 rgba(0,0,0,.1)', 
//   },
//   darkTheme: {
//     bodyBkg: '#111',
//     bodyTextColor: '#ccc',
//     footerBkg: '#000',
//     footerTextColor: '#ccc',

//     formBorder: '1px solid #aaa3',

//     formBkg: '#000!important',
//     formBoxShadow: '0 0.5rem 1.2rem rgba(124,124,124,.2)',
//     formTransition: '.3s all ease',

//     formBkgHover: '#bbb1!important',
//     formBoxShadowHover: '2px 5px 10px 0 rgba(0,0,0,.2)',
//     formTransitionHover: '.2s all ease',

//     formHeaderDescColor: '#00c087', // Contrast Checker:https://webaim.org/resources/contrastchecker/
    
//     inputBkg: '#ccc2',
//     inputBorder: '1px solid #555',
//     inputBorderHover: '#1274e0',

//     resultBoxShadow: 'inset 0 0 5px 0 rgba(0,0,0,.3)',
//   },
//   moonTheme: {
//     bodyBkg: '#01263f',
//     bodyTextColor: '#AABCCB',
//     footerBkg: '#012137',
//     footerTextColor: '#AABCCB',

//     formBorder: '1px solid #18365b',

//     formBkg: '#012137!important',
//     // formBoxShadow: '0 2px 2px 0 rgba(0,0,0,.1)',
//     formBoxShadow: '0 0.5rem 1.2rem rgba(4,76,124,.2)',
//     formTransition: '.3s all ease',

//     formBkgHover: '#bbb1!important',
//     formBoxShadowHover: '2px 5px 10px 0 rgba(0,0,0,.3)',
//     // formBoxShadowHover: '0 0.5rem 1.2rem rgba(4,76,124,.2)',
//     formTransitionHover: '.2s all ease',

//     formHeaderDescColor: '#00D696', // Contrast Checker:https://webaim.org/resources/contrastchecker/
    
//     inputBkg: '#ccd2',
//     inputBorder: '1px solid #256',
//     inputBorderHover: '#2ac',

//     resultBoxShadow: 'inset 0 0 5px 0 rgba(0,0,0,.3)',
//     // resultBoxShadow: 'inset 0 0 10px #ddd;',
//   }
// };
// export default themes;