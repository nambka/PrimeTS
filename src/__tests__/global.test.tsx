import renderer from 'react-test-renderer';
import GlobalStyle from '../service/theme/global';
import { lightTheme, darkTheme, moonTheme} from '../service/theme';

describe('GlobalStyle: ', () => {
  it('renders with a LIGHT theme', () => {
    renderer.create(<GlobalStyle theme={lightTheme} />);
    expect(document.head).toMatchSnapshot();
  });
  it('renders with a DARK theme', () => {
    renderer.create(<GlobalStyle theme={darkTheme} />);
    expect(document.head).toMatchSnapshot();
  });
  it('renders with a MOON theme', () => {
    renderer.create(<GlobalStyle theme={moonTheme} />);
    expect(document.head).toMatchSnapshot();
  });

})