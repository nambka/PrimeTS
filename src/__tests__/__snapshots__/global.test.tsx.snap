// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalStyle:  renders with a DARK theme 1`] = `
<head>
  <style
    data-styled="active"
    data-styled-version="5.2.1"
  >
    *,*::after,*::before{box-sizing:border-box;}
    body{background:#F7F8F9;color:#353637;font-family:BlinkMacSystemFont,-apple-system,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;-webkit-transition:all 0.25s linear;transition:all 0.25s linear;}
    footer{background:linear-gradient(150deg,#19a0ff,#2d1582);color:#fff;}
    form{background-color:#fff!important;border:1px solid #eaecef22;box-shadow:0 5px 30px 0 rgba(0,0,0,.2);-webkit-transition:.3s all ease;transition:.3s all ease;}
    form:hover{background-color:#fff;border:1px solid #eaecef22;box-shadow:0 3px 10px 0 rgba(204,204,204,.8);-webkit-transition:.2s all ease;transition:.2s all ease;}
    form .nambk-header-desc{color:#006647;}
    input[type=number]{color:#353637;background-color:#fff3;border:1px solid #E6E8EA;}
    input[type=number]:focus{border-color:#F0B90B;}
    #result{box-shadow:inset 0 0 5px 0 rgba(0,0,0,.1);}
    *,*::after,*::before{box-sizing:border-box;}
    body{background:#111;color:#ccc;font-family:BlinkMacSystemFont,-apple-system,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;-webkit-transition:all 0.25s linear;transition:all 0.25s linear;}
    footer{background:#000;color:#ccc;}
    form{background-color:#000!important;border:1px solid #aaa3;box-shadow:0 0.5rem 1.2rem rgba(124,124,124,.2);-webkit-transition:.3s all ease;transition:.3s all ease;}
    form:hover{background-color:#bbb1!important;border:1px solid #aaa3;box-shadow:2px 5px 10px 0 rgba(0,0,0,.2);-webkit-transition:.2s all ease;transition:.2s all ease;}
    form .nambk-header-desc{color:#00c087;}
    input[type=number]{color:#ccc;background-color:#ccc2;border:1px solid #555;}
    input[type=number]:focus{border-color:#1274e0;}
    #result{box-shadow:inset 0 0 5px 0 rgba(0,0,0,.3);}
  </style>
</head>
`;

exports[`GlobalStyle:  renders with a LIGHT theme 1`] = `
<head>
  <style
    data-styled="active"
    data-styled-version="5.2.1"
  >
    *,*::after,*::before{box-sizing:border-box;}
    body{background:#F7F8F9;color:#353637;font-family:BlinkMacSystemFont,-apple-system,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;-webkit-transition:all 0.25s linear;transition:all 0.25s linear;}
    footer{background:linear-gradient(150deg,#19a0ff,#2d1582);color:#fff;}
    form{background-color:#fff!important;border:1px solid #eaecef22;box-shadow:0 5px 30px 0 rgba(0,0,0,.2);-webkit-transition:.3s all ease;transition:.3s all ease;}
    form:hover{background-color:#fff;border:1px solid #eaecef22;box-shadow:0 3px 10px 0 rgba(204,204,204,.8);-webkit-transition:.2s all ease;transition:.2s all ease;}
    form .nambk-header-desc{color:#006647;}
    input[type=number]{color:#353637;background-color:#fff3;border:1px solid #E6E8EA;}
    input[type=number]:focus{border-color:#F0B90B;}
    #result{box-shadow:inset 0 0 5px 0 rgba(0,0,0,.1);}
  </style>
</head>
`;

exports[`GlobalStyle:  renders with a MOON theme 1`] = `
<head>
  <style
    data-styled="active"
    data-styled-version="5.2.1"
  >
    *,*::after,*::before{box-sizing:border-box;}
    body{background:#F7F8F9;color:#353637;font-family:BlinkMacSystemFont,-apple-system,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;-webkit-transition:all 0.25s linear;transition:all 0.25s linear;}
    footer{background:linear-gradient(150deg,#19a0ff,#2d1582);color:#fff;}
    form{background-color:#fff!important;border:1px solid #eaecef22;box-shadow:0 5px 30px 0 rgba(0,0,0,.2);-webkit-transition:.3s all ease;transition:.3s all ease;}
    form:hover{background-color:#fff;border:1px solid #eaecef22;box-shadow:0 3px 10px 0 rgba(204,204,204,.8);-webkit-transition:.2s all ease;transition:.2s all ease;}
    form .nambk-header-desc{color:#006647;}
    input[type=number]{color:#353637;background-color:#fff3;border:1px solid #E6E8EA;}
    input[type=number]:focus{border-color:#F0B90B;}
    #result{box-shadow:inset 0 0 5px 0 rgba(0,0,0,.1);}
    *,*::after,*::before{box-sizing:border-box;}
    body{background:#111;color:#ccc;font-family:BlinkMacSystemFont,-apple-system,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;-webkit-transition:all 0.25s linear;transition:all 0.25s linear;}
    footer{background:#000;color:#ccc;}
    form{background-color:#000!important;border:1px solid #aaa3;box-shadow:0 0.5rem 1.2rem rgba(124,124,124,.2);-webkit-transition:.3s all ease;transition:.3s all ease;}
    form:hover{background-color:#bbb1!important;border:1px solid #aaa3;box-shadow:2px 5px 10px 0 rgba(0,0,0,.2);-webkit-transition:.2s all ease;transition:.2s all ease;}
    form .nambk-header-desc{color:#00c087;}
    input[type=number]{color:#ccc;background-color:#ccc2;border:1px solid #555;}
    input[type=number]:focus{border-color:#1274e0;}
    #result{box-shadow:inset 0 0 5px 0 rgba(0,0,0,.3);}
    *,*::after,*::before{box-sizing:border-box;}
    body{background:#01263f;color:#AABCCB;font-family:BlinkMacSystemFont,-apple-system,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;-webkit-transition:all 0.25s linear;transition:all 0.25s linear;}
    footer{background:#012137;color:#AABCCB;}
    form{background-color:#012137!important;border:1px solid #18365b;box-shadow:0 0.5rem 1.2rem rgba(4,76,124,.2);-webkit-transition:.3s all ease;transition:.3s all ease;}
    form:hover{background-color:#bbb1!important;border:1px solid #18365b;box-shadow:2px 5px 10px 0 rgba(0,0,0,.3);-webkit-transition:.2s all ease;transition:.2s all ease;}
    form .nambk-header-desc{color:#00D696;}
    input[type=number]{color:#AABCCB;background-color:#ccd2;border:1px solid #256;}
    input[type=number]:focus{border-color:#2ac;}
    #result{box-shadow:inset 0 0 5px 0 rgba(0,0,0,.3);}
  </style>
</head>
`;
